Simple ROS driver for UNITREE A1

sunbo57123 opened the pull request:
A simple implementation of ROS driver for unitree A1.

Commit:
feat: add basic files

Commit:
feat: add Walk function

receive geometry_msg/Twist msgs and convert it into UNITREE highcmd,
then send the high cmd to unitree A1.

Commit:
fix: udp loog funcs exit error.

UNITREE_LEGGED_SDK::LoopFunc should be implemented in the function which
runs ROS nodes.

Commit:
refactor: rename "unitree" to "unitree a1" or "a1"

This ROS driver is currently tested on unitree A1 robot, not for all unitree
products.

Commit:
style: reformat code

crystaldust commented on .gitignore:
```diff
@@ -0,0 +1,6 @@
+install/*
+.idea/*
+transfer.sh
+*/.idea/*
```
Duplicated, a single ".idea/*" or ".idea/" would be enough.

sunbo57123 replied:
line 4 and line 6? They are different paths. 😕

crystaldust commented on a1_driver/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+set(EXTRA_LIBS ${LEGGED_SDK_NAME} lcm)
+
+add_executable(a1_main src/main.cpp
+        src/main.cpp
```
Duplicated reference

crystaldust commented on a1_driver/src/A1_ros.cpp:
```diff
@@ -0,0 +1,26 @@
+            "cmd_vel",
+            10,
+            [this](geometry_msgs::msg::Twist::UniquePtr msg) {
+                std::cout << "received: speed = " << msg->linear.x << std::endl;
```
Considering replacing all std::cout with rclcpp's logging system.

crystaldust commented on a1_driver/src/A1_ros.cpp:
```diff
@@ -0,0 +1,26 @@
+    );
+//    A1_node->create_subscription<...>();
+    UNITREE_LEGGED_SDK::InitEnvironment();
+    float dt = 0.002f;
```
It would be friendly to make the duration configurable for the developers.

sunbo57123 replied:
I prefer setting a default number, this is not necessary for user to config, and the influence is not markable. If you think it is important, I would change it.

crystaldust replied:
Giving a default value makes sense. But developers might run the code in different hardwares, if the value is too high or too low for specific hardware, we should at least give an option to change it.

sunbo57123 replied:
This is the driver for unitree A1, the hardware wouldn't have other choices.

crystaldust commented on a1_driver/src/A1_ros.cpp:
```diff
@@ -0,0 +1,26 @@
+    UNITREE_LEGGED_SDK::InitEnvironment();
+    float dt = 0.002f;
+    UNITREE_LEGGED_SDK::LoopFunc loop_udpRecv("udp_recv", dt, 3, boost::bind(&A1Wrapper::UDPRecv, &wrapper));
+    UNITREE_LEGGED_SDK::LoopFunc loop_udpSend("udp_send", dt, 3, boost::bind(&A1Wrapper::UDPSend, &wrapper));
```
I'm not very sure of the second parameter's meaning, described as 'int bindCPU'. It's OK to just keep it 3 here, but we should make the meaning clear in the future.

crystaldust commented on a1_driver/src/A1_ros.cpp:
```diff
@@ -0,0 +1,26 @@
+int A1ROS::node_init(int argc, char *argv[]) {
+    auto A1_node = rclcpp::Node::make_shared(this->node_name);
+    auto A1_sub = A1_node->create_subscription<geometry_msgs::msg::Twist>(
+            "cmd_vel",
```
Better add a prefix "unitree_a1/" before the "cmd_vel", which makes the scope clear.

crystaldust commented on a1_driver/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+
+link_directories($ENV{UNITREE_LEGGED_SDK_PATH}/lib)
+string(CONCAT LEGGED_SDK_NAME libunitree_legged_sdk_$ENV{UNITREE_PLATFORM}.so)
+set(EXTRA_LIBS ${LEGGED_SDK_NAME} lcm)
```
The driver doesn't rely on lcm anymore. DDS will handle real-time messaging.

sunbo57123 replied:
LCM is used in unitree sdk building, the process of colcon build would have error : "undefined reference to "lcm_XX" " without lcm.

crystaldust commented on a1_driver/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+ament_target_dependencies(a1_main
+        ${dependencies}
+        )
+target_link_libraries(a1_main
```
Suggestion: if the variable list is not very long, it's OK to make it one line.

crystaldust commented on a1_driver/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+add_executable(msg_pub examples/msg_pub.cpp)
+ament_target_dependencies(msg_pub
+        ${dependencies}
+        )
```
Suggestion: the closing brackets should obey the same rule, making it in the beginning looks good.

crystaldust commented on a1_driver/examples/msg_pub.cpp:
```diff
@@ -0,0 +1,41 @@
+class PubNode : public rclcpp::Node {
+public:
+    PubNode() : Node("pub_node"), count_(0) {
+        walk_pub = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
```
Just remind: the topic should be synced to "unitree_a1/cmd_vel" with the subscriptions updates.

crystaldust commented on a1_driver/include/unitree_a1_ros2/A1_wrapper.h:
```diff
@@ -0,0 +1,40 @@
+#define UNITREE_DRIVER_A1_WRAPPER_H
+
+#include <iostream>
+#include "time.h"
```
- We should hide the internal sleep behaviors within the implementation, so time.h should be included in the A1_wrapper.cpp file.
- It's better to use <time.h> instead of "time.h" to clearly claim that time.h is a standard library.

Commit:
refactor: fix problems in #4

1. change "std::out" in callback function to "RCLCPP_INFO"
2. remove duplicate reference in CMAKE file
3. change the place of "#include <time.h>"
4. change topic name from "cmd_vel" to "unitree_a1/cmd_vel"
5. change project name more specific

Commit:
refactor: rename the folder under include

rename "include/unitree_a1_ros2" to "include/a1_driver"
